### ISSUE TITLE
migration / upgrade from previously working installation is overloading (?) MQTT 

### ISSUE THREAD

**Issue by rsteininger** [DaveMDS/node-red-contrib-tasmota#27]
Hi there, 
Today I had upgraded to the latest version of node red (1.3.3 from maybe 1.3.1 or so). I had 38 instances of tasmota working which included sensors, buttons, etc. 
However, when updating to the latest tasmota node version, all hell broke lose on my mqtt server (mosquito). It took me about 2 hours to pinpoint the problem to tasmota. Only once I pointed the tasmota mqtt settings to a non existing mqtt server, the rest of my home automation went back to normal. Every couple of seconds I had "burst" of over 200 mqtt msg/ sec, completely parallising my entire home automation system. 
Experimenting with a single tasmota node and the correct mqtt settings did not yield any working results any longer either while all other nodes using mqtt work fine. 
Not sure what exactly happend but all of my 38 node-red-contrib-tasmota nodes are down (while e.g. node-red-contrib-sonoff-tasmota-enhanced or mqtt receive work perfectly fine). 
As my tasmota node configurations are partially complex, deleting and replacing them all would be a very bad option for me. 

Here a screenshot showing the debug output when I "freshly" create a new node-red-contrib-tasmota node and new MQTT settings, while node-red-contrib-sonoff-tasmota-enhanced works perfectly fine. 
![image](https://user-images.githubusercontent.com/5363948/115974452-bbb57780-a554-11eb-9447-1c7098f03958.png)

Not sure what it is but something seems "incompatible" with MQTT now! 

Is there anything I can do to provide additional information for further troubleshooting or any other (manual) tricks to restore normal operation? So far I was really happy with this tasmota node and it would be a shame if I had to replace it all. 

THANKS
Robert

**Comment by DaveMDS** [comment]
Hi, first of all sorry for the inconvenience, indeed I have done some hard changes to the mqtt connection that was really needed to go for a stable release.

Please show me the configuration of the 2 nodes (and the 2 mqtt broker configs) for the image you posted.

In the image I can barely see a "user already connected" error, is that related to the Tasmota node?

For upgrading your already deployed Tasmota nodes you must go on each node config, choose again the mqtt broker and save the config. This must be done for ALL nodes, also if the configuration "seems" already correct. Then restart nodered.

Also give a look at #26, maybe there are interesting info for you

**Comment by rsteininger** [comment]
Thank you Dave for your quick reply! I am aware that this is all your personal time creating and maintaining these!

I am familiar with configuring mqtt config nodes as other nodes have a similar approach. Here a screenshot of the new tasmota mqtt configuration:
![image](https://user-images.githubusercontent.com/5363948/115985347-68b6e100-a5a3-11eb-9999-49b8aff1acb1.png)

And yes, I had edited all 38 instances and selected the new mqtt config, tested "modified nodes" and also did a "full" redeploy, restarted node red and even rebooted my pi 4.

I think my symptoms are similar to the ones described in the post #26 but I didn't really see any resolution beside deleting them all and starting over from scratch. 

What I have done currently is configured all 38 nodes to point to a non existing mqtt broker so that the rest of my home automation came back online (some 70 devices), as I said, my mqtt server (mosquitto) did go bonkers once all 38 tasmota nodes were connected. I had the symptom that the last tasmota node I configured did sometime come online briefly and I was able to operate the same. BUT, if this was the second tasmota node a changed, it would "break" the previous one that might have worked before. Very odd ... I can't be sure what the exact cause or effect was but as stated, my mqtt server came down on it's knees. I took my only other source of frequent mqtt messages (zigbee2mqtt) offline as I had upgraded to a new version at the same time but to no avail. But, as mentioned above, only misconfiguring the mqtt IP address for tasmota brought my system back online and functional again. 

For testing I created a singel new "node-red-contrib-tasmota" node with a second tasmota mqtt configuration (this time the correct one) but I don't seem to get it working (anymore). What's interesting is, that even the deploy of the single new node takes "ages" and it ends with this message: 25/04/2021, 08:56:15node: mqtt198
msg : string[36]
"Error stopping node: Close timed out"
![image](https://user-images.githubusercontent.com/5363948/115985579-5db08080-a5a4-11eb-85f8-aab9771613ed.png)

When I do the same with "node-red-contrib-sonoff-tasmota-enhanced" the socket01 works right away. 

I have re-entered the mqtt credentials several times (copy past from lastpass as I have a longish password) so, I don't think it is a typo issue. 

The one thing I am very reluctant to do at this time is to delete all 38 nodes, uninstall tasmota, restart node red and start all over ... that would be a "weekend long" job ... 

I had noticed a good while back (don't ask me what version that was) that sometimes the tasmota nodes "hang up". This usually happens during setup / configuration with the device going on and offline. If that happens, in those instances, the tasmota node (very very sporadically) gets stuck in the offline mode, despite that the device is online and reporting to my mqtt broker. WHEN that happens, I had to delete the node create a new one and reattach all my connections and I do wonder, if now this is such a situation ... I am NOT a linux guru but can follow instructions if there is anything I can provide you with for a further analysis, logs, screenshots, configuration details, etc. 

If there is a way to e.g. manually remove tasmota from the command line, restart node red and reinstall tasmota (without having to delete my 38 devices) I would be willing to give that a go ... 

Many, many thanks! 
Robert

**Comment by rsteininger** [comment]
Interesting, I am back to "cycling", I did once more what I described above, created a new tasmota node and configured it that same as the one stuck in offline (now below the other sonoff node):
![image](https://user-images.githubusercontent.com/5363948/115986447-2217b580-a5a8-11eb-8ebe-15658e66520e.png)
And now it starts "cycling" again. 
The broker flashes back and forth between "Broker connected" and "Broker disconnected" producing these messages in the debug windo: 
25/04/2021, 09:25:43node: mqtt198
msg : error
"Error: read ECONNRESET"
![image](https://user-images.githubusercontent.com/5363948/115986521-76bb3080-a5a8-11eb-8493-2a9b26f9febe.png)
I can "control" the node via the inject nodes BUT there is a big delay (3 to 5 seconds) and the tasmota node continues to report disconnect issues. 
Any idea on how to further troubleshoot? 
THANKS

**Comment by DaveMDS** [comment]
Hi, it seems your mqtt broker is not accepting the connection, don't you see some interesting message in the mosquito logs?

I suspect your broker is not accepting 2 connections with the same Username and/or same ClientID. Keep in mind that the tasmota nodes establish a different connection to your broker (previously it was sharing the connection with other mqtt nodes) , so if you have a standard mqtt node and a Tasmota node then two different connections are performed to the broker.

Try to change the Username in the Tasmota nodes config so that it is different than the Username in other mqtt nodes. And look at the mosquito logs, you should find the cause of the disconnection there

**Comment by rsteininger** [comment]
Hi Dave,
Thanks! I have created a new user in my mosquitto mqtt server and at first, it seemed to "solve" my issue. However, as I went slowly through my 38 previously working tasmota instances, which I switched over only one by one, the ones that had been online and working once more, were sporadically dropping off again, it seems, impacted by the more nodes I had converted to the new mqtt broker (with the new user). Some nodes I could not get online at all, despite deleting the tasmota node and creating it from scratch, while the sonoff node works consistently fine. They would once more show all kinds of conditions, from offline to Broker connecting, frozen in state "OFF" etc. No way of telling in what state they would end up. I could not get any shelly 2.5 (dual channel relay) back online and as a rule of thumb, where I use a sensor along with a switching node with the same ID (e.g. socket with power monitoring), either the switch or the sensor or both would not come online and work properly (again, each of these work perfectly fine with mqtt receive or the sonoff node, except for the dual channel relay where I don't have a work-around yet). 
But what really got me is that the sockets that were at first fixed later did not work anymore once I had reconfigured other tasmota devices. Typically, when I try and configure a node that gives trouble, at some point I would once more get the mqtt broker error message in debug, but keep in mind, the mqtt broker settings have not changed since changing over the first tasmota device to the mqtt broker with the new user ID. I can't at present restart node red, which may or may not fix some issues. I cannot explain what is going on but something is not right. 
I have no experience with checking the mosquitto log for errors as in my initial attempt I got flooded with normal operations messages drowning out any warnings or errors ... I will have to spend some time with that but can't at the moment due to work. 
I have (created) work arounds for most essential tasmota nodes except for the dual channel relays which is a problem at the moment (my problem, I know). 
Anything else you could think of to try would be greatly appreciated! 
thanks,
Robert

**Comment by rsteininger** [comment]
Hi Dave, 
I have restarted node red this morning but no change. I still have a fair number of nodes off line (some that have already worked on the new user name / password) and restarting node red did not change which ones worked. 
Happy to provide further details / analysis if you guide me what you need to see.
Thanks,

**Comment by DaveMDS** [comment]
Hi, you really need to look at the mosquitto log, it should tell what is going wrong

**Comment by DaveMDS** [comment]
Another idea: 
make sure all your tasmota nodes are using the same tasmote-mqtt-broker node, 
and also make sure you have a single tasmote-mqtt-broker configured in NodeRed

**Comment by rsteininger** [comment]
Thank you Dave! 
With regards to the two brokers - when the new tasmota nodes had caused havoc in my home automation system, I had changed the setting to point to a non existing mqtt server IP which restored the rest of my system. Then I created a second mqtt broker pointing to the correct mqtt server. After that, I moved over the nodes one by one and as I said, initially, the nodes with the new mqtt broker setting worked fine. That deteriorated however the more nodes I had moved over. Nevertheless, even when having moved over ALL RELEVANT nodes to the new setting, the second, old, incorrect mqtt broker was still there and some not active (test) tasmota nodes were still pointing to that. If I read your instructions from this AM, I should not have done that and that might have caused an issue? (This would be an issue in the long run as I have some special purpose tasmota nodes that use a different public broker to bridge communication to an offsite installation ... 

Yesterday, I had activated full logging in my mosquitto server. When analysing the logs I could not readily identify issues created by the "new" tasmota nodes. I found the successful broker communication initiation (as it is the only one using the new mqtt user name) and only one entry about a duplication of nodes with same ID (not an error per se but an indication that communication was established already. The message was correct in my test setup, there was a duplication of a node with the same ID). But otherwise no warnings or errors that I could find. There were however within minutes several thousand entries in the log file (log all = all actives, not just warnings and errors) and I wouldn't have been able to analyse them all (I probably checked close to 500 lines and filtered all entries for warnings and errors). It did help me however to find an issue with another part of my home automation system but that was independent from the tasmota nodes...

I had a perfectly fine working setup with 38 tasmota nodes & sensors grown over the last year or so with regular extensions. My home automation was brought down to its knees and had limited functionality over the last two days. I have now spent hours and hours since to recover. So yesterday afternoon I made the painful decision to switch my solution from your tasmota nodes to a combination of sonoff and mqtt in and out nodes. Not as elegant and sleek as your nodes but functional for my purpose. That transfer was finished last night and all is working again as it was before upgrading. 

As I said, I cannot be sure what has caused these issues, whether something in my particular installation (combination of things?), a legacy (?) configuration issue with the tasmota nodes or something incompatible in the current tasmota software version. However, I have a mixture of homematic, zigbee, tasmota nodes and web api interfaces running (including weather forecast, integration of our home ev charger AND our BMW i3 talking to each other for charging). So, the setup is becoming rather ... extensive. Overall it is performing well. Besides regular maintenance and upgrades every couple of weeks, the system runs 24/7, implying that there are no system stability issues on a larger scale I am aware of. 

It pained me to abandon an elegant solution like your tasmota nodes but with my limited knowledge, there is only that much I can do (before my wife kicks me out as she can't turn on the light in the utility room ... ;-) )

Nevertheless, thanks for your help! 
All the best, 
Robert
